tensorflow_probability/python/mcmc/replica_exchange_mc.py

```diff
@@ -0,0 +1,365 @@
+  """
+  def default_exchange_proposed_fn_(n_replica, seed=None):
+    seed = distributions_util.gen_new_seed(seed, "default_exchange_proposed_fn")
+    u = tf.random_uniform([], seed=seed)
+    exist = u > freq
+
+    seed = distributions_util.gen_new_seed(seed, "default_exchange_proposed_fn")
+    i = tf.random_uniform([], minval=0, maxval=2, dtype=tf.int32, seed=seed)
+    a = tf.range(i, n_replica - 1, 2)
+    exchange_proposed = tf.transpose(tf.concat(([a], [a + 1]), axis=0))
+    exchange_proposed_n = tf.to_int32(tf.floor(
+        (tf.to_int32(n_replica) - i) / 2))
+
+    exchange_proposed = tf.cond(
+        exist, lambda: tf.to_int32([]), lambda: exchange_proposed)
+    exchange_proposed_n = tf.cond(exist, lambda: 0, lambda: exchange_proposed_n)
+    return exchange_proposed, exchange_proposed_n
+  return default_exchange_proposed_fn_
+
+
+class ReplicaExchangeMC(kernel_base.TransitionKernel):
+  """Runs one step of the Replica Exchange Monte Carlo.
+
+  [Replica Exchange Monte Carlo](
+  https://en.wikipedia.org/wiki/Parallel_tempering) is a Markov chain
+  Monte Carlo (MCMC) algorithm that is also known as Parallel Tempering. This
+  algorithm performs multiple sampling with different temperatures in parallel,
+  and exchange those samplings according to the Metropolis-Hastings criterion.
+  By using the sampling result of high temperature, sampling with less influence
+  of the local solution becomes possible.
+
+  #### Examples
+
+  ```python
+  import tensorflow as tf
+  import tensorflow_probability as tfp
+  import numpy as np
+  import matplotlib.pyplot as plt
+
+  tfd = tf.contrib.distributions
+
+  # Tuning acceptance rates:
+  dtype = np.float32
+  num_warmup_iter = 1000
+  num_chain_iter = 1000
+
+  x = tf.get_variable(name='x', initializer=np.zeros(2, dtype=dtype))
+
+  # Target distribution is Mixture Normal.
+  target = tfd.MixtureSameFamily(
+      mixture_distribution=tfd.Categorical(probs=[0.5, 0.5]),
+      components_distribution=tfd.MultivariateNormalDiag(
+          loc=[[-5., -5], [5., 5.]],
+          scale_identity_multiplier=[1., 1.]))
+
+  # Initialize the ReplicaExchangeMC sampler.
+  remc = tfp.mcmc.ReplicaExchangeMC(
+      target_log_prob_fn=target.log_prob,
+      inverse_temperatures=tf.pow(10., tf.linspace(0., -2., 5)),
+      replica_kernel_class=tfp.mcmc.HamiltonianMonteCarlo,
+      step_size=0.5,
+      num_leapfrog_steps=3)
+
+  # One iteration of the ReplicaExchangeMC
+  init_results = remc.bootstrap_results(x)
+  next_x, other_results = remc.one_step(
+      current_state=x,
+      previous_kernel_results=init_results)
+
+  x_update = x.assign(next_x)
+  replica_update = [init_results.replica_states[i].assign(
+      other_results.replica_states[i]) for i in range(remc.n_replica)]
+
+  warmup = tf.group([x_update, replica_update])
+
+  init = tf.global_variables_initializer()
+
+  with tf.Session() as sess:
+    sess.run(init)
+    # Warm up the sampler
+    for _ in range(num_warmup_iter):
```

Comment:
This example is great as is. However it might be nice to teach users about our sample_chain driver, which is muuuch more efficient that repeated calls.  Eg,

    sampled_states, kernel_results = tfp.mcmc.sample_chain(
        num_results=num_chain_iter,
        init_state=tf.zeros([num_chain_iter, 2]),
        num_burnin_steps=num_warmup_iter)

This approach also saves you the trouble of using `tf.Variables`. This is good because its a simpler example and (hopefully) dramatically more efficient.

Reply:
Thanks! I'll add new example of sample_chain.

==============================================
I tried using `sample_chain`, but I got an error.

```python
dtype = np.float32

target = tfd.Normal(loc=dtype(0), scale=dtype(1))

samples, _ = tfp.mcmc.sample_chain(
  num_results=1000,
  current_state=dtype(1),
  kernel=tfp.mcmc.RandomWalkMetropolis(
     target.log_prob,
     seed=42),
  num_burnin_steps=500,
  parallel_iterations=1)  # For determinism.
```

Error is this.
```
TypeError: The two structures don't have the same sequence type. First structure has type <class 'tensorflow_probability.python.mcmc.random_walk_metropolis.UncalibratedRandomWalkResults'>, while second structure has type <class 'list'>.
```

I think this cause is that a kernel result before `one_step` and this after `one_step` are different.

```python
with tf.Session() as sess:
  target = tfd.Normal(loc=dtype(0), scale=dtype(1))
  kernel = tfp.mcmc.RandomWalkMetropolis(
          target.log_prob,
          seed=42)
  
  cs = dtype(1)
  pkr = kernel.bootstrap_results(cs)
  
  print(pkr.accepted_results)
  cs, pkr = kernel.one_step(cs, pkr)
  print(pkr.accepted_results)
```

This code prints this.
```
UncalibratedRandomWalkResults(log_acceptance_correction=<tf.Tensor 'mh_bootstrap_results/rwm_bootstrap_results/zeros_like:0' shape=() dtype=float32>, target_log_prob=<tf.Tensor 'mh_bootstrap_results/rwm_bootstrap_results/Normal/log_prob/sub:0' shape=() dtype=float32>)

[<tf.Tensor 'mh_one_step/choose_inner_results/Select:0' shape=() dtype=float32>, <tf.Tensor 'mh_one_step/choose_inner_results/Select_1:0' shape=() dtype=float32>]
```

I rewrote `is_namedtuple_like` like this. Then I got no error.
```python:tensorflow_probability/python/mcmc/util.py
def is_namedtuple_like(x):
  """Helper which returns `True` if input is `collections.namedtuple`-like."""
  return (isinstance(x, tuple) or
          isinstance(getattr(x, '__dict__', None), collections.Mapping) and
          getattr(x, '_fields', None) is not None)
```